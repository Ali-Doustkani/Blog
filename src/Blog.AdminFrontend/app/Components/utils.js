const ask = fn => id => {
   if (confirm('Are you sure?')) {
      fn(id)
   }
}

const emitChange = props => args => {
   props.onChange({
      id: props.id,
      [props.name]: args.target.value
   })
}

const renderClassNames = props => {
   const errors = props[props.name + 'Errors']
   const className = props.className
   if (errors && errors.length) {
      return className ? className + ' incorrect' : 'incorrect'
   }
   return className
}

export { ask, emitChange, renderClassNames }
