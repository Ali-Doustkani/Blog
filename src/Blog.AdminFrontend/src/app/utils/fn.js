const compose = (...functions) => args => functions.reduceRight((arg, fn) => fn(arg), args)

const dict = (...pairs) => key => {
   try {
      return pairs.find(pair => pair[0] === key)[1]
   } catch {
      throw new Error('Invalid key for pair')
   }
}

const ifElse = (predicate, onTrue, onFalse) => (state, action) =>
   predicate(action) ? onTrue(state, action) : onFalse(state, action)

export { compose, dict, ifElse }
