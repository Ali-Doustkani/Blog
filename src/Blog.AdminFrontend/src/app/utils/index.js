export { map } from './serverErrorMapper'
export { emptyValidator, richtextEmptyValidator } from './validations'
