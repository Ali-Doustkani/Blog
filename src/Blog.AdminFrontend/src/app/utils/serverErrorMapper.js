const makeMessage = error => {
   const propName = error.path[error.path.length - 1]
   switch (error.error) {
      case 'isRequired':
         return `${propName} is required`
      case 'isEmpty':
         return `${propName} can not be empty`
      default:
         return `something is wrong with ${propName}`
   }
}

const read = error => ({ type: 2, message: makeMessage(error) })

const mapPath = (error, obj) =>
   error.path.reduce(
      (parent, prop, index) =>
         error.path.length - 1 === index ? parent[prop + 'Errors'] : parent[prop],
      obj
   )

const map = (errors, obj) =>
   errors.forEach(error => {
      mapPath(error, obj).push(read(error))
   })

export { map }
