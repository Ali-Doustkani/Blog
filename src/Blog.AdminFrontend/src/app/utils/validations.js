const isWhitespace = input => input.match(/^\s*$/) !== null

const isEmpty = input => !input || isWhitespace(input)

const isRichtextEmpty = input => {
   const result = /<p.*?>(?<inner>.*?)<\/p>/.exec(input)
   return result ? isEmpty(result.groups.inner) : isEmpty(input)
}

const getDuplicate = config => {
   for (const item of config.values) {
      return config.values.find(isDuplicate(item, config))
   }
}

const isDuplicate = (item1, config) => item2 =>
   propsNotEmpty(item1, item2, config) &&
   compareProps(item1, item2, config) &&
   item1.id !== item2.id

const propsNotEmpty = (item1, item2, config) =>
   Array.isArray(config.prop)
      ? config.prop.every(p => !isEmpty(item1[p]) && !isEmpty(item2[p]))
      : !isEmpty(item1[config.prop]) && !isEmpty(item2[config.prop])

const compareProps = (item1, item2, config) =>
   Array.isArray(config.prop)
      ? config.prop.every(p => item1[p] === item2[p])
      : item1[config.prop] === item2[config.prop]

const hasOverlap = (obj1, obj2) => {
   const startA = new Date(obj1.startDate)
   const endA = new Date(obj1.endDate)
   const startB = new Date(obj2.startDate)
   const endB = new Date(obj2.endDate)

   return startA < endB && endA > startB
}

const getOverlap = objs => {
   for (const item1 of objs) {
      for (const item2 of objs) {
         if (item1.id !== item2.id && hasOverlap(item1, item2)) {
            return [item2, item1]
         }
      }
   }
   return [undefined, undefined]
}

const isArrayEmpty = array => !array && array.length

export { isEmpty, isRichtextEmpty, isArrayEmpty, getDuplicate, hasOverlap, getOverlap }
