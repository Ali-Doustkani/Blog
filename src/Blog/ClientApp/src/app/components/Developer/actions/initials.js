const STATUS = {
   LOADING: 1,
   SAVING: 2,
   IDLE: 3,
   STOPPED: 4
}

const initialState = {
   status: STATUS.LOADING,
   errorMessage: '',
   summary: '',
   summaryErrors: [],
   skills: '',
   skillsErrors: [],
   experiences: [],
   experiencesErrors: [],
   sideProjects: [],
   sideProjectErrors: []
}

export { initialState, STATUS }
