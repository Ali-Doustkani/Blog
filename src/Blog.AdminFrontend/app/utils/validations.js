import { compose } from './fn'

const isWhitespace = input => input.match(/^\s*$/) !== null

const isEmpty = input => !input || isWhitespace(input)

const isRichtextEmpty = input => {
   const result = /<p.*?>(?<inner>.*?)<\/p>/.exec(input)
   return result ? isEmpty(result.groups.inner) : isEmpty(input)
}

const message = msg => result => (result ? msg : undefined)

const requiredMessage = field => message(`${field} is required`)

const errorListGenerator = errors => message => {
   const result = !errors || !errors.length ? [] : errors.filter(x => x.type === 2)
   if (message) {
      result.push({ type: 1, message })
   }
   return result
}

const emptyValidator = (field, errors) =>
   compose(
      errorListGenerator(errors),
      requiredMessage(field),
      isEmpty
   )

const richtextEmptyValidator = (field, errors) =>
   compose(
      errorListGenerator(errors),
      requiredMessage(field),
      isRichtextEmpty
   )

export { emptyValidator, richtextEmptyValidator }
