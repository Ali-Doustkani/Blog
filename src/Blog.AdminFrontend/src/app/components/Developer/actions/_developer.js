import { STATUS } from './initials'

const updateDeveloper = (state, action) => ({ ...state, ...action.change })

const updateIds = (state, action) => ({
   ...state,
   experiences: state.experiences.map(setId(action.data.experiences)),
   status: STATUS.IDLE
})

const setId = ids => (exp, index) => ({ ...exp, id: ids[index] })

export { updateDeveloper, updateIds }
