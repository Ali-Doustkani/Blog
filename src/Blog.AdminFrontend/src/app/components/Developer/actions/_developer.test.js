import { updateDeveloper, showErrors, updateIds, prepareForSave } from './_developer'
import { initialState, STATUS } from './initials'
import { richtextEmptyValidator, emptyValidator } from '../../../utils'

jest.mock('./_serverErrorWriter')
jest.mock('../../../utils')

describe('updating', () => {
   it('updates fields', () => {
      emptyValidator.mockImplementation(() => () => [])
      richtextEmptyValidator.mockImplementation(() => () => [])
      let state = updateDeveloper(initialState, { change: { summary: 'summary' } })
      state = updateDeveloper(state, { change: { skills: 'C#' } })

      expect(state).toMatchSnapshot()
   })
})

describe('updating ids', () => {
   it('sets ids by order', () => {
      const state = {
         experiences: [
            {
               id: 'a',
               company: 'Parmis'
            },
            {
               id: 'b',
               company: 'Lodgify'
            }
         ]
      }
      const newState = updateIds(state, { data: { experiences: [1, 2] } })

      expect(newState.experiences[0].id).toBe(1)
      expect(newState.experiences[1].id).toBe(2)
   })

   it('sets status', () => {
      const state = {
         status: STATUS.LOADING,
         experiences: []
      }
      const newState = updateIds(state, { data: { experiences: [] } })

      expect(newState.status).toBe(STATUS.IDLE)
   })
})

describe('showErrors', () => {
   it('sets status to idle', () => {
      const initial = {
         status: STATUS.LOADING
      }
      const state = showErrors(initial, { data: null })

      expect(state.status).toBe(STATUS.IDLE)
   })
})

describe('prepare for saving', () => {
   it('removes all errors with server type', () => {
      const init = {
         summaryErrors: [{ type: 1, message: 'err1' }, { type: 2, message: 'err2' }],
         skillsErrors: [{ type: 1, message: 'err1' }, { type: 2, message: 'err2' }],
         experiencesErrors: [{ type: 1, message: 'err1' }, { type: 2, message: 'err2' }],
         experiences: [
            {
               companyErrors: [{ type: 1, message: 'err1' }, { type: 2, message: 'err2' }],
               positionErrors: [{ type: 1, message: 'err1' }, { type: 2, message: 'err2' }],
               startDateErrors: [{ type: 1, message: 'err1' }, { type: 2, message: 'err2' }],
               endDateErrors: [{ type: 1, message: 'err1' }, { type: 2, message: 'err2' }],
               contentErrors: [{ type: 1, message: 'err1' }, { type: 2, message: 'err2' }]
            }
         ]
      }

      const newState = prepareForSave(init)

      const errors = [{ type: 1, message: 'err1' }]
      expect(newState.summaryErrors).toEqual(errors)
      expect(newState.skillsErrors).toEqual(errors)
      expect(newState.experiencesErrors).toEqual(errors)
      expect(newState.experiences[0].companyErrors).toEqual(errors)
      expect(newState.experiences[0].positionErrors).toEqual(errors)
      expect(newState.experiences[0].startDateErrors).toEqual(errors)
      expect(newState.experiences[0].endDateErrors).toEqual(errors)
      expect(newState.experiences[0].contentErrors).toEqual(errors)
   })
})
