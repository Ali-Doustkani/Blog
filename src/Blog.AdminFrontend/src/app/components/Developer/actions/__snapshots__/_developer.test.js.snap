// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updating updates fields 1`] = `
Object {
  "errorMessage": "",
  "experiences": Array [],
  "experiencesErrors": Array [],
  "skills": "C#",
  "skillsErrors": Array [],
  "status": 1,
  "summary": "summary",
  "summaryErrors": Array [],
}
`;
